File: python/firefox_bidi_client/bidi_page.py
import asyncio
import base64
import json
from pathlib import Path
from typing import Any, Optional

from .bidi_connection import BiDiConnection
from .logger import log_debug


def _unwrap_remote_value(remote: Any) -> Any:
    if not remote or not isinstance(remote, dict):
        return remote
    t = remote.get("type")
    if t in ("string", "number", "boolean"):
        return remote.get("value")
    if t in ("null", "undefined"):
        return None
    if t == "array":
        return [_unwrap_remote_value(v) for v in (remote.get("value") or [])]
    if t == "object":
        val = remote.get("value")
        if isinstance(val, list):
            return {_unwrap_remote_value(k): _unwrap_remote_value(v) for k, v in val}
        return val
    return remote


def _serialize_local_value(value: Any) -> Any:
    if value is None:
        return {"type": "null"}
    if isinstance(value, bool):
        return {"type": "boolean", "value": value}
    if isinstance(value, (int, float)):
        return {"type": "number", "value": value}
    if isinstance(value, str):
        return {"type": "string", "value": value}
    if isinstance(value, list):
        return {"type": "array", "value": [_serialize_local_value(v) for v in value]}
    if isinstance(value, dict):
        return {
            "type": "object",
            "value": [
                [{"type": "string", "value": k}, _serialize_local_value(v)]
                for k, v in value.items()
            ],
        }
    raise TypeError(f"Cannot serialize {type(value).__name__} to BiDi LocalValue")


class BiDiPage:
    def __init__(self, connection: BiDiConnection, context_id: str):
        self._connection = connection
        self._context_id = context_id

    @property
    def id(self) -> str:
        return self._context_id

    # ── Navigation ──────────────────────────────────────────────────────────────

    async def goto(self, url: str, wait: str = "complete") -> None:
        await self._connection.send_command("browsingContext.navigate", {
            "context": self._context_id,
            "url": url,
            "wait": wait,
        })
        log_debug(f"Navigated to {url}")

    async def navigate(self, url: str, wait: str = "complete") -> None:
        return await self.goto(url, wait)

    # ── Script evaluation ────────────────────────────────────────────────────────

    async def evaluate(self, expression: str, arg: Any = None) -> Any:
        """
        Evaluate a JS expression or function declaration string in the page context.

            await page.evaluate('document.title')
            await page.evaluate('(x) => x * 2', 21)
            await page.evaluate('() => document.querySelectorAll("a").length')
        """
        stripped = expression.strip()
        is_function = "=>" in stripped or stripped.startswith("function")

        if is_function:
            result = await self._connection.send_command("script.callFunction", {
                "functionDeclaration": expression,
                "target": {"context": self._context_id},
                "awaitPromise": True,
                "arguments": [_serialize_local_value(arg)] if arg is not None else [],
            })
        else:
            result = await self._connection.send_command("script.evaluate", {
                "expression": expression,
                "target": {"context": self._context_id},
                "awaitPromise": True,
            })

        if result.get("type") == "exception":
            raise RuntimeError(
                result.get("exceptionDetails", {}).get("text", "Script exception")
            )
        return _unwrap_remote_value(result.get("result"))

    # ── Interaction ──────────────────────────────────────────────────────────────

    async def perform_actions(self, actions: list) -> None:
        """Send one or more BiDi input source actions and release when done.
        Accepts the raw actions array from the input.performActions spec."""
        await self._connection.send_command("input.performActions", {
            "context": self._context_id,
            "actions": actions,
        })
        await self._connection.send_command("input.releaseActions", {"context": self._context_id})

    async def click(self, selector: str) -> None:
        located = await self._connection.send_command("browsingContext.locateNodes", {
            "context": self._context_id,
            "locator": {"type": "css", "value": selector},
            "maxNodeCount": 1,
        })
        nodes = located.get("nodes", [])
        if not nodes:
            raise RuntimeError(f"Element not found: {selector}")

        element = nodes[0]
        # x/y 0,0 = element center; BiDi scrolls the element into view automatically
        await self.perform_actions([{
            "type": "pointer",
            "id": "mouse1",
            "actions": [
                {
                    "type": "pointerMove",
                    "x": 0,
                    "y": 0,
                    "origin": {"type": "element", "element": {"sharedId": element["sharedId"]}},
                },
                {"type": "pointerDown", "button": 0},
                {"type": "pointerUp", "button": 0},
            ],
        }])
        log_debug(f'Clicked "{selector}"')

    async def type(self, selector: str, text: str) -> None:
        """Click a selector to focus it, then dispatch key down/up pairs for each character.
        For special keys use their Unicode code point (e.g. '\\uE007' for Enter)."""
        await self.click(selector)
        await self.perform_actions([{
            "type": "key",
            "id": "keyboard",
            "actions": [
                action
                for char in text
                for action in (
                    {"type": "keyDown", "value": char},
                    {"type": "keyUp", "value": char},
                )
            ],
        }])
        log_debug(f'Typed {len(text)} characters into "{selector}"')

    # ── Waiting ──────────────────────────────────────────────────────────────────

    async def wait_for_function(
        self,
        expression: str,
        timeout: float = 5.0,
        interval: float = 0.1,
    ) -> None:
        """Poll until the JS expression returns truthy, or timeout (seconds)."""
        loop = asyncio.get_event_loop()
        deadline = loop.time() + timeout
        while loop.time() < deadline:
            try:
                if await self.evaluate(expression):
                    return
            except Exception:
                pass
            await asyncio.sleep(interval)
        raise TimeoutError(f"wait_for_function timed out after {timeout}s")

    async def wait_for_selector(self, selector: str, timeout: float = 5.0) -> None:
        return await self.wait_for_function(
            f"document.querySelector({json.dumps(selector)}) !== null",
            timeout=timeout,
        )

    async def wait_for(
        self,
        expression: str,
        timeout: float = 5.0,
        interval: float = 0.1,
    ) -> None:
        return await self.wait_for_function(expression, timeout=timeout, interval=interval)

    # ── Page info ────────────────────────────────────────────────────────────────

    async def title(self) -> str:
        return await self.evaluate("document.title")

    async def url(self) -> str:
        return await self.evaluate("location.href")

    # ── Screenshot ───────────────────────────────────────────────────────────────

    async def screenshot(self, path: Optional[str] = None) -> bytes:
        result = await self._connection.send_command("browsingContext.captureScreenshot", {
            "context": self._context_id,
        })
        data = base64.b64decode(result["data"])
        if path:
            Path(path).write_bytes(data)
            log_debug(f'Screenshot saved to "{path}"')
        return data

File: src/bidi-page.ts
import { writeFileSync } from 'node:fs';
import { BiDiConnection } from './bidi-connection.js';
import { logDebug } from './logger.js';

export interface NavigateOptions {
  timeout?: number;
}

export interface WaitForOptions {
  timeout?: number;
  interval?: number;
}

export interface EvaluateOptions {
  timeout?: number;
}

export interface ClickOptions {
  timeout?: number;
}

export interface PerformActionsOptions {
  timeout?: number;
}

export interface ScreenshotOptions {
  path?: string;
  timeout?: number;
}

// ── BiDi remote value ────────────────────────────────────────────────────────

function unwrapRemoteValue(remote: any): any {
  if (!remote || typeof remote !== 'object') return remote;
  switch (remote.type) {
    case 'string':
    case 'number':
    case 'boolean':
      return remote.value;
    case 'null':
      return null;
    case 'undefined':
      return undefined;
    case 'array':
      return (remote.value ?? []).map(unwrapRemoteValue);
    case 'object':
      if (Array.isArray(remote.value)) {
        return Object.fromEntries(
          remote.value.map(([k, v]: [any, any]) => [unwrapRemoteValue(k), unwrapRemoteValue(v)])
        );
      }
      return remote.value;
    default:
      return remote;
  }
}

// Serialize a JS value into a BiDi LocalValue for script.callFunction arguments
function serializeLocalValue(value: any): any {
  if (value === null)      return { type: 'null' };
  if (value === undefined) return { type: 'undefined' };
  if (typeof value === 'string')  return { type: 'string',  value };
  if (typeof value === 'number')  return { type: 'number',  value };
  if (typeof value === 'boolean') return { type: 'boolean', value };
  if (Array.isArray(value)) {
    return { type: 'array', value: value.map(serializeLocalValue) };
  }
  if (typeof value === 'object') {
    return {
      type: 'object',
      value: Object.entries(value).map(([k, v]) => [
        { type: 'string', value: k },
        serializeLocalValue(v),
      ]),
    };
  }
  throw new Error(`Cannot serialize value of type "${typeof value}" to BiDi LocalValue`);
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function withTimeout<T>(promise: Promise<T>, ms: number | undefined, label: string): Promise<T> {
  if (ms === undefined) return promise;
  return Promise.race([
    promise,
    sleep(ms).then((): never => { throw new Error(`${label} timed out after ${ms}ms`); }),
  ]);
}

// ── BiDiPage ─────────────────────────────────────────────────────────────────

export class BiDiPage {
  constructor(
    private readonly connection: BiDiConnection,
    private readonly contextId: string
  ) {}

  get id(): string {
    return this.contextId;
  }

  // ── Navigation ─────────────────────────────────────────────────────────────

  async goto(url: string, options: NavigateOptions = {}): Promise<void> {
    const { timeout = 5000 } = options;

    await Promise.race([
      this.connection.sendCommand('browsingContext.navigate', {
        context: this.contextId,
        url,
        wait: 'complete',
      }),
      sleep(timeout).then(() => logDebug(`Navigation to ${url} timed out after ${timeout}ms`)),
    ]);

    logDebug(`Navigated to ${url}`);
  }

  /** @alias goto */
  async navigate(url: string, options?: NavigateOptions): Promise<void> {
    return this.goto(url, options);
  }

  // ── Script evaluation ──────────────────────────────────────────────────────

  /**
   * Evaluate a JS expression string or function in the page context.
   * Matches Playwright's page.evaluate() signature.
   *
   *   await page.evaluate('document.title')
   *   await page.evaluate(() => document.title)
   *   await page.evaluate(selector => document.querySelector(selector).textContent, '#heading')
   */
  async evaluate(expression: string | ((...args: any[]) => any), arg?: any, options: EvaluateOptions = {}): Promise<any> {
    const cmd = typeof expression === 'function'
      ? this.connection.sendCommand('script.callFunction', {
          functionDeclaration: expression.toString(),
          target: { context: this.contextId },
          awaitPromise: true,
          arguments: arg !== undefined ? [serializeLocalValue(arg)] : [],
        })
      : this.connection.sendCommand('script.evaluate', {
          expression,
          target: { context: this.contextId },
          awaitPromise: true,
        });

    const result = await withTimeout(cmd, options.timeout, 'evaluate');

    if (result.type === 'exception') {
      throw new Error(result.exceptionDetails?.text ?? 'Script exception');
    }
    return unwrapRemoteValue(result.result);
  }

  // ── Interaction ────────────────────────────────────────────────────────────

  /**
   * Send one or more BiDi input source actions and release when done.
   * Accepts the raw `actions` array from the input.performActions spec.
   */
  async performActions(actions: any[], options: PerformActionsOptions = {}): Promise<void> {
    await withTimeout(
      this.connection.sendCommand('input.performActions', { context: this.contextId, actions }),
      options.timeout,
      'performActions',
    );
    await this.connection.sendCommand('input.releaseActions', { context: this.contextId });
  }

  async click(selector: string, options: ClickOptions = {}): Promise<void> {
    const located = await withTimeout(
      this.connection.sendCommand('browsingContext.locateNodes', {
        context: this.contextId,
        locator: { type: 'css', value: selector },
        maxNodeCount: 1,
      }),
      options.timeout,
      `locateNodes(${selector})`,
    );

    if (!located.nodes?.length) {
      throw new Error(`Element not found: ${selector}`);
    }

    const element = located.nodes[0];

    // x/y 0,0 = element center; BiDi scrolls the element into view automatically
    await this.performActions([{
      type: 'pointer',
      id: 'mouse1',
      actions: [
        { type: 'pointerMove', x: 0, y: 0, origin: { type: 'element', element: { sharedId: element.sharedId } } },
        { type: 'pointerDown', button: 0 },
        { type: 'pointerUp', button: 0 },
      ],
    }], options);
    logDebug(`Clicked "${selector}"`);
  }

  /**
   * Click a selector to focus it, then dispatch key down/up pairs for each character.
   * For special keys use their Unicode code point (e.g. '' for Enter).
   */
  async type(selector: string, text: string): Promise<void> {
    await this.click(selector);
    await this.performActions([{
      type: 'key',
      id: 'keyboard',
      actions: text.split('').flatMap(char => [
        { type: 'keyDown', value: char },
        { type: 'keyUp', value: char },
      ]),
    }]);
    logDebug(`Typed ${text.length} characters into "${selector}"`);
  }

  // ── Waiting ────────────────────────────────────────────────────────────────

  /**
   * Poll until the expression/function returns truthy, or timeout.
   * Matches Playwright's page.waitForFunction().
   */
  async waitForFunction(
    expression: string | ((...args: any[]) => any),
    options: WaitForOptions = {}
  ): Promise<void> {
    const { timeout = 5000, interval = 100 } = options;
    const deadline = Date.now() + timeout;

    while (Date.now() < deadline) {
      try {
        if (await this.evaluate(expression)) return;
      } catch {
        // element may not exist yet — keep polling
      }
      await sleep(interval);
    }

    throw new Error(`waitForFunction timed out after ${timeout}ms`);
  }

  /** Wait until selector matches an element in the DOM. */
  async waitForSelector(selector: string, options: WaitForOptions = {}): Promise<void> {
    return this.waitForFunction(
      `document.querySelector(${JSON.stringify(selector)}) !== null`,
      options
    );
  }

  /** @alias waitForFunction */
  async waitFor(
    expression: string | ((...args: any[]) => any),
    options?: WaitForOptions
  ): Promise<void> {
    return this.waitForFunction(expression, options);
  }

  // ── Page info ──────────────────────────────────────────────────────────────

  async title(): Promise<string> {
    return this.evaluate('document.title');
  }

  async url(): Promise<string> {
    return this.evaluate('location.href');
  }

  // ── Screenshot ─────────────────────────────────────────────────────────────

  /**
   * Capture a screenshot. Returns a Buffer (matching Playwright's return type).
   * Pass { path } to also save to disk.
   */
  async screenshot(options: ScreenshotOptions = {}): Promise<Buffer> {
    const result = await withTimeout(
      this.connection.sendCommand('browsingContext.captureScreenshot', { context: this.contextId }),
      options.timeout,
      'screenshot',
    );

    const buffer = Buffer.from(result.data as string, 'base64');

    if (options.path) {
      writeFileSync(options.path, buffer);
      logDebug(`Screenshot saved to "${options.path}"`);
    }

    return buffer;
  }
}
